File: Tree/binary_search_tree/insertion/contructioncpp/construction.cpp
#include<iostream>
#define COUNT 10
using namespace std;
class node{
	public:
	int data;
	node *left;
	node *right;
	node(int key){
		data = key;
		left = right = NULL;
	}
};
class bst{
	public:
	node * root;
	bst(){
		root = NULL;
	}
	bst(int key){
		root = new node(key);
	}
	void Insert(node *r,int value){
		if(r == NULL){
			r = new node(value);
			r->left = NULL;
			r->right = NULL;
			return;
		}
		else{
			if(value>r->data){
				Insert(r->right,value);
				return;
			}
			else{
				Insert(r->left,value);
				return;
			}
		}
	}
};
void print2DUtil(node *root, int space){
    if (root == NULL)
        return;
    space += COUNT;
    print2DUtil(root->right, space);
    cout<<endl;
    for (int i = COUNT; i < space; i++)
        cout<<" ";
    cout<<root->data<<"\n";
    print2DUtil(root->left, space);
}
void print2D(node *root){
    print2DUtil(root, 0);
}

int main(){
	bst *tree = new bst();
	int nodes;
	cout<<"how many nodes?"<<endl;
	cin>>nodes;
	int value;
	int i;
	for(i=0;i<nodes;i++){
		cin>>value;
		tree->Insert(tree->root,value);
	}
	print2DUtil(tree->root,2);
	return 0;
}

File: Tree/binary_search_tree/insertion/constructionjava/construction.java
    class node{
	    int data;
	    node left;
	    node right;
	    node(int key){
		    data = key;
		    left = right = null;
	    }
    }

    class bst{
	    node  root;
	    node  temp;
	    node  last;
	    int flag = 0;
	    bst(){
		    root =  null;
	    }
	    bst(int key){
		    root = new node(key);
	    }
	    void Insert(node r,int value){
		    temp = r;///saving root
		    if(temp == null){
			    if(root == null){
				    root = new node(value);
				    root.data = value;
				    return;
			    }
			    temp = new node(value);
			    temp.data = value;
			    return;
		    }
		    else{
			    if(value > temp.data){
				    Insert(temp.right,value);
				    return;
			    }
			    else{
				    Insert(temp.left,value);
				    return;
			    }
		    }
	    }
    }

    class construction{
	    static void in_order(node root){
		    if(root == null){
			    return;
		    }
		    in_order(root.left);
		    System.out.println(root.data+" ");
		    in_order(root.right);
	    }
	    public static void main(String[] args){
		    bst tree = new bst();
		    System.out.println(tree.root);
		    tree.Insert(tree.root,45);
		    System.out.println(tree.root);
		    tree.Insert(tree.root,39);
		    System.out.println(tree.root);
		    tree.Insert(tree.root,12);
		    System.out.println(tree.root);
		    tree.Insert(tree.root,59);
		    test.in_order(tree.root);
	    }
    }
